torchx/c_src/torchx.cpp

```diff
@@ -0,0 +1,583 @@
+
+  // Clone here to copy data from blob, which will be GCed.
+  TENSOR(at::clone(at::from_blob(blob.data, shape, type)));
+}
+
+NIF(type)
+{
+  TENSOR_PARAM(0, t);
+
+  std::string type_name = type2string(t->scalar_type());
+
+  if (!type_name.empty())
+    return nx::nif::ok(env, enif_make_atom(env, type_name.c_str()));
+  else
+    return nx::nif::error(env, "Could not determine tensor type.");
+}
+
+NIF(device)
+{
+  TENSOR_PARAM(0, t);
+
+  at::optional<at::Device> device = at::device_of(*t);
+
+  if (device.has_value())
+    return nx::nif::ok(env, nx::nif::make(env, device.value().str().c_str()));
+  else
+    return nx::nif::error(env, "Could not determine tensor device.");
+}
+
+NIF(nbytes)
+{
+  TENSOR_PARAM(0, t);
+
+  return nx::nif::ok(env, enif_make_int64(env, t->nbytes()));
+}
+
+NIF(split)
+{
+  TENSOR_PARAM(0, t);
+  PARAM(1, int64_t, batch_size);
+
+  TENSOR_LIST(at::split(*t, batch_size));
+}
+
+NIF(to_blob)
+{
+  ERL_NIF_TERM result;
+  TENSOR_PARAM(0, t);
+  size_t byte_size = t->nbytes();
+
+  if (argc == 2)
+  {
+    PARAM(1, int64_t, limit);
+    byte_size = limit * t->itemsize();
+  }
+
+  void *result_data = (void *)enif_make_new_binary(env, byte_size, &result);
+  memcpy(result_data, t->data_ptr(), byte_size);
+
+  return result;
+}
+
+NIF(scalar_tensor)
+{
+  PARAM(0, double, scalar);
+  TYPE_PARAM(1, type);
+
+  TENSOR(at::scalar_tensor(scalar, type));
+}
+
+NIF(randint)
+{
+  PARAM(0, int64_t, min);
+  PARAM(1, int64_t, max);
+  SHAPE_PARAM(2, shape);
+  TYPE_PARAM(3, type);
+
+  TENSOR(at::randint(min, max, shape, type));
+}
+
+NIF(rand)
+{
+  PARAM(0, double, min);
+  PARAM(1, double, max);
+  SHAPE_PARAM(2, shape);
+  TYPE_PARAM(3, type);
+
+  TENSOR(min + at::rand(shape, type) * (max - min));
+}
+
+NIF(normal)
+{
+  PARAM(0, double, mean);
+  PARAM(1, double, std);
+  SHAPE_PARAM(2, shape);
+
+  TENSOR(at::normal(mean, std, shape));
+}
+
+NIF(arange)
+{
+  PARAM(0, int64_t, start);
+  PARAM(1, int64_t, end);
+  PARAM(2, int64_t, step);
+  TYPE_PARAM(3, type);
+
+  if (argc == 5)
+  {
+    SHAPE_PARAM(4, shape);
+    TENSOR(at::reshape(at::arange((double)start, (double)end, (double)step, type), shape));
+  }
+  else
+  {
+    TENSOR(at::arange((double)start, (double)end, (double)step, type));
+  }
+}
+
+NIF(iota)
+{
+  SHAPE_PARAM(0, shape);
+  PARAM(1, int, axis);
+  TYPE_PARAM(2, type);
+
+  try
+  {
+    // gets the size of iota
+    int64_t dim = shape[axis];
+
+    // build the iota in one dimension
+    at::Tensor aten = at::arange(0.0, (double)dim, 10.0, type);
+
+    std::vector<int64_t> rshape(shape.size());
+    std::fill(rshape.begin(), rshape.end(), 1);
+    rshape[axis] = dim;
+
+    aten = at::reshape(aten, rshape);
+
+    at::Tensor result = at::broadcast_to(aten, shape).clone();
+
+    return nx::nif::ok(env, create_tensor_resource(env, result));
+  }
+  catch (c10::Error error)
+  {
+    return nx::nif::error(env, error.msg().c_str());
+  }
+}
```

Comment:
Personally, I would prefer for this to be implemented in Elixir and keep the NIF to be simple wrappers around libtorch. After all, it is easier to maintain Elixir code than C++ one, but there is no problem if you would prefer to keep this function as is.

Reply:
Done!